sw/device/lib/dif/dif_flash_ctrl.h

```diff
@@ -0,0 +1,807 @@
+
+/**
+ * Get the size of this region in number of pages.
+ *
+ * @param flash_ctrl FLASH CTRL device to query.
+ * @param region The region in question.
+ * @param size_out Out-parameter, the number of pages in this `region`.
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` or `enable_out` is null, or if
+ * `region` is invalid, `kDifFlashCtrlOk` otherwise.
+ */
+dif_flash_ctrl_result_t dif_flash_ctrl_region_size_get(
+    const dif_flash_ctrl_t *flash_ctrl, uint32_t region, uint32_t *size_out);
+
+/**
+ * Set the partition for this region.
+ *
+ * This may only be done if region configuration has not been locked.
+ *
+ * @param flash_ctrl FLASH CTRL device to configure.
+ * @param region The region to operate on.
+ * @param part The partition to set for this `region`.
+ * @return `kDifFlashCtrlRegionBadArg` if `flash_ctrl` is null or `region` is
+ * invalid, `kDifFlashCtrlRegionBadAccess` if region configuration has been
+ * locked, and `kDifFlashCtrlRegionOk` otherwise.
+ */
+dif_flash_ctrl_region_result_t dif_flash_ctrl_region_partition_set(
+    const dif_flash_ctrl_t *flash_ctrl, uint32_t region,
+    dif_flash_ctrl_part_sel_t part);
+
+/**
+ * Get the partition this region refers to.
+ *
+ * @param flash_ctrl FLASH CTRL device to query.
+ * @param region The region in question.
+ * @param part_out Out-parameter, the active partition for this `region`.
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` or `enable_out` is null, or if
+ * `region` is invalid, `kDifFlashCtrlOk` otherwise.
+ */
+dif_flash_ctrl_result_t dif_flash_ctrl_region_partition_get(
+    const dif_flash_ctrl_t *flash_ctrl, uint32_t region,
+    dif_flash_ctrl_part_sel_t *part_out);
+
+/**
+ * Lock memory region configuration until the device is reset.
+ *
+ * This will prevent any further configuration of memory regions until device
+ * reset. Future calls to region configuration functions will return
+ * `kDifFlashCtrlRegionBadAccess`.
+ *
+ * @param flash_ctrl FLASH CTRL device to lock region configuration on.
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` is null, `kDifFlashCtrlOk`
+ * otherwise.
+ */
+dif_flash_ctrl_result_t dif_flash_ctrl_regions_config_disable(
+    const dif_flash_ctrl_t *flash_ctrl);
+
+/**
+ * Query the state of the memory region configuration registers.
+ *
+ * This function checks if memory region configuration is still enabled or if
+ * has been locked. Once locked, region configuration cannot be enabled again,
+ * and all calls to region configuration functions will return
+ * `kDifFlashCtrlRegionBadAccess` until the device is restarted.
+ *
+ * @param flash_ctrl FLASH CTRL device to check the lock state for.
+ * @param access_out Out-parameter, the current state of memory region
+ * configuration access.
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` or `access_out` is null,
+ * `kDifFlashCtrlOk` otherwise.
+ */
+dif_flash_ctrl_result_t dif_flash_ctrl_region_config_check(
```

Comment:
I was being careful with the language here to not use the word "lock", since in the hw docs "enabled" indicates configuration is enabled, rather than the lock being enabled. `Region register write enable. Once set to 0, it can longer be configured to 1`

Avoiding the word "lock" in the function name allows us to reuse the generic `dif_flash_ctrl_enable` struct as the out-parameter and remain consistent with the hw docs.